ports/espressif/common-hal/sdioio/SDCard.c

```diff
@@ -0,0 +1,223 @@
+    if (common_hal_sdioio_sdcard_deinited(self)) {
+        raise_deinited_error();
+    }
+}
+
+static int check_pins(const mcu_pin_obj_t *clock, const mcu_pin_obj_t *command, const uint8_t num_data, const mcu_pin_obj_t **data) {
+    if (CONFIG_SOC_SDMMC_USE_GPIO_MATRIX) {
+        // ESP32-S3 and P4 can use any pin for any SDMMC func in either slot
+        // Default to SLOT_1 for SD cards
+        ESP_LOGI(TAG, "Using chip with CONFIG_SOC_SDMMC_USE_GPIO_MATRIX");
+        return SDMMC_HOST_SLOT_1;
+    }
+    if (command->number == GPIO_NUM_11 && clock->number == GPIO_NUM_6 && data[0]->number == GPIO_NUM_7) {
+        // Might be slot 0
+        if (num_data == 1 || (num_data == 4 && data[1]->number == GPIO_NUM_8 && data[2]->number == GPIO_NUM_9 && data[3]->number == GPIO_NUM_10)) {
+            return SDMMC_HOST_SLOT_0;
+        }
+    } else if (command->number == GPIO_NUM_15 && clock->number == GPIO_NUM_14 && data[0]->number == 2) {
+        // Might be slot 1
+        if (num_data == 1 || (num_data == 4 && data[1]->number == GPIO_NUM_4 && data[2]->number == GPIO_NUM_12 && data[3]->number == GPIO_NUM_13)) {
+            return SDMMC_HOST_SLOT_1;
+        }
+    }
+    return -1;
+}
+
+void common_hal_sdioio_sdcard_construct(sdioio_sdcard_obj_t *self,
+    const mcu_pin_obj_t *clock, const mcu_pin_obj_t *command,
+    uint8_t num_data, const mcu_pin_obj_t **data, uint32_t frequency) {
+    if (num_data != 4 && num_data != 1) {
+        mp_raise_ValueError(MP_ERROR_TEXT("SDIO: cards have either 1 or 4 data lines"));
+    }
+
+    self->num_data = num_data;
+    int sd_slot = check_pins(clock, command, num_data, data);
+    if (sd_slot != SDMMC_HOST_SLOT_0 && sd_slot != SDMMC_HOST_SLOT_1) {
+        // Bad pin combo
+        raise_ValueError_invalid_pins();
+    }
+
+    if (frequency > 40000000) {
+        // Higher than max 40Mhz frequency
+        mp_raise_ValueError(MP_ERROR_TEXT("SDIO: requested frequency out of range"));
+    }
+
+    ESP_LOGI(TAG, "Using slot %d", sd_slot);
+    self->slot = (uint8_t)sd_slot;
+    esp_err_t err = ESP_OK;
+
+    sdmmc_host_t host = SDMMC_HOST_DEFAULT();
+    host.max_freq_khz = frequency / 1000;
+
+    sdmmc_slot_config_t slot_config = SDMMC_SLOT_CONFIG_DEFAULT();
+    slot_config.width = 1;
+    slot_config.clk = clock->number;
+    claim_pin(clock);
+    self->clock = clock->number;
+    slot_config.cmd = command->number;
+    claim_pin(command);
+    self->command = command->number;
+    slot_config.d0 = data[0]->number;
+    self->data[0] = data[0]->number;
+    claim_pin(data[0]);
+    if (num_data == 4) {
+        slot_config.width = 4;
+        slot_config.d1 = data[1]->number;
+        claim_pin(data[1]);
+        self->data[1] = data[1]->number;
+        slot_config.d2 = data[2]->number;
+        claim_pin(data[2]);
+        self->data[2] = data[2]->number;
+        slot_config.d3 = data[3]->number;
+        claim_pin(data[3]);
```

Comment:
Move all of these pin claims after the inits below that can throw errors. That way pins can be reused if the init fails.

Reply:
Thanks, done.